### ISSUE TITLE
open GL - graphics overlaping

### ISSUE THREAD

**Issue by fintogive** [TASVideos/desmume#248]
![mario kart ds moon__1946](https://user-images.githubusercontent.com/32790406/51296302-6763da80-19d0-11e9-9a99-de761ee6bf72.png)

 i hope is a good example of the issue above.  basically when  a texture is mapped to a object  and its touching  the very edge of its usable space,  in open GL mode only the bottom part over laps to the top of the 2d model.  and probably the same if the texture is touching left to right or visa versa.

**Comment by rogerman** [comment]
@fintogive, I'm not getting you.

Are you saying that the black horizontal line over the rock-looking thingy is your problem?

I also noticed that you are running a non-native resolution. Does this issue also occur when you're running the native resolution?

**Comment by zeromus** [comment]
This is something a game is likely to do on accident, and it may not be noticeable on the original hardware, or maybe it is and nobody ever noticed it. More examples are required.

**Comment by zeromus** [comment]
1 texel offset in 3d rendering in desmume is not noteworthy. It's been like this since the beginning of time and will likely always be this way for some cases in 3d.

**Comment by rogerman** [comment]
Well... if the issue is just a simple off-by-one texture sampling error, then this certainly won't be fixed until our 3D engine is completely reworked to use purely integers instead of floats.

When that happens, then it MIGHT be possible to fix this kind of error in OpenGL by carefully manipulating the texture sampling coordinates in the shaders. But again, we need to rework the underlying 3D engine first.

**Comment by fintogive** [comment]
> 
> 
> 1 texel offset in 3d rendering in desmume is not noteworthy. It's been like this since the beginning of time and will likely always be this way for some cases in opengl renderer

Actually it work as intended in the last stable build 0.9.11  and open gl was pretty buggy then.    so if we can be optimistic perhaps it will be fixed before then end of time.

**Comment by fintogive** [comment]
Thanks for the information Rogerman hopefully it can be fixed one day.

**Comment by zeromus** [comment]
The 1 pixel details of texturing have never been guaranteed in desmume. It may break or fix randomly as various things are changed. You need to stop obsessing over the moon rock. If there are a lot of newly broken things, that is interesting. One random newly broken 1px thing out of the millions of things that could be broken across all games.. is not interesting. If there was a real bug in this rock then there would be real bugs in a lot of places

**Comment by fintogive** [comment]
> 
> 
> The 1 pixel details of texturing have never been guaranteed in desmume. It may break or fix randomly as various things are changed. You need to stop obsessing over the moon rock. If there are a lot of newly broken things, that is interesting. One random newly broken 1px thing out of the millions of things that could be broken across all games.. is not interesting. If there was a real bug in this rock then there would be real bugs in a lot of places

Im not obsessing over anything.  its a bug in the program  this would happen with any textures with the same characteristics in any ds game  its not just about the moon rock...  im just pointing it out in hopes that one day it will get fixed im in no rush to have it fixes. i cant help that its not a mayor bug or multiple bugs or an interesting bug that im reporting.  

The fact that it was working at one point means it got broken by other changes.  personally i like to be positive about it being fixed at one point in time. i cant help that it doesn't interest you.
i appreciate what you do, when you do it but a bug is a bug, big or small.  and i will report it when i see it.  its that a problem?  Your majesty?

**Comment by zeromus** [comment]
If it will happen with any textures with the same characteristics in any DS game, then show me one. I think the characteristics are very rare: an object placed peculiarly, transformed peculiarly, with that exact visual characteristic (material at edge and transparent on the other side), with the wrap mode set to wrap instead of clamp for no apparent reason (probably on accident by the artist). You should be getting materials with that visual characterstic commonly (trees, rock, fences, etc.) but they won't ordinarily have the wrong wrap mode set. (It's also possible that texture is in an atlas and opengl is sampling too far, like happens with letters in many games, but I doubt it)

**Comment by fintogive** [comment]
Ok ill hunt one down and show you when i get home it but i know its desmume and not the ds game like i said it work as intended in hardware and softrasterizer

**Comment by Jules-A** [comment]
You'd be better trying to find a smaller regression range, sometime between 2011 and now doesn't really help much. Check if it worked in a build from a year ago, if so try 6 months ago etc.

**Comment by zeromus** [comment]
No, youre better off finding more examples. The "regression" will be some minor adjustment which cannot logically be changed and WOULD not be to fix one moon rock which I contend is bugged in the original game to begin with, even if it isn't visible on hardware.

**Comment by fintogive** [comment]
> 
> 
> You'd be better trying to find a smaller regression range, sometime between 2011 and now doesn't really help much. Check if it worked in a build from a year ago, if so try 6 months ago etc.

the issue has existed for over a year.  i'll have to go back to the sourceforge builds

**Comment by fintogive** [comment]
build 5397  off Orphis  works perfectly.  if the builds didn't expire id hunt it back to when it started showing up.

**Comment by rogerman** [comment]
@fintogive, we have nightly builds going all the way back to r5505. Just scroll down to the bottom of the [desmume.org/download](http://desmume.org/download) page to see it.

If r5505 works for you, then you should have a decently good record of nightly builds on our downloads page to start your regression search.

**Comment by fintogive** [comment]
> 
> 
> @fintogive, we have nightly builds going all the way back to r5505. Just scroll down to the bottom of the [desmume.org/download](http://desmume.org/download) page to see it.
> 
> If r5505 works for you, then you should have a decently good record of nightly builds on our downloads page to start your regression search.

that was what i was going to do next but nevermind scratch that. its the anti aliasing.  i forgotten i had turn that on it was set to x2  and at x2 its at its worst  the higher it is the less noticeable it is.  disabling it fixes the issue.

**Comment by fintogive** [comment]
Alright.  Sorry for wasting your time. i thought it was a legit issue.

**Comment by zeromus** [comment]
i wonder if, as a hack, we could change any textures set to to use texcoords exactly on the 0 or 1 edges of a texture, to use clamp instead of wrap. like, the rule is, any time the texture couldn't possibly wrap, due to choice of texture coords, fix the choice of wrapping mode to be clamp, so that we dont accidentally wrap due to 3d renderer standard inaccuracies like this.

**Comment by fintogive** [comment]
its worth a shot.

**Comment by rogerman** [comment]
I don't think it's the tex coords -- I think it's the position coords. MSAA shouldn't have any effect on tex coords, but MSAA can certainly change position coords. Probably what's happening is that MSAA is ever-so-slightly increasing the size of the polygon, and if the texture is set to wrap, then that's what the texture is going to do in within the extra polygon space.

Well... in any case, it should be possible to introduce such a hack in the OpenGL renderer. We would have to analyze the tex coords of each polygon in BeginRender() and flag such polygons to force clamped texture sampling. Then during the rendering loop, use the "forced texture clamp flag" to determine whether to force the texture sampling mode. Shouldn't be too difficult.

**Comment by zeromus** [comment]
MSAA causes texels you didnt intend to sample to get sampled indeed. that's the classic bug (or at least one of them). If you meant to sample across texels row at y=0, it will actually sample from y=-1. It's not that it changes the texcoords but it makes stuff render that you simply didn't ask for. awful hack, it is.

**Comment by rogerman** [comment]
Hack implemented. Please see commit 02cd950 for details.

**Comment by Jules-A** [comment]
> 
> 
> Hack implemented. Please see commit [02cd950](https://github.com/TASVideos/desmume/commit/02cd950b02ea5767eb195f1eb4615f96be005dca) for details.

Doesn't fix any of the similar issues in the Pokemon games.
